### PR TITLE
Refactor dynamic requirements of reduced events.

### PR DESCRIPTION
**Pull request by riga**
This PR refactors the dynamic dependencies we have in place between `MergeReduceEvents` and everything below it: `UniteColumns`, `ProduceColumns`, `PrepareMLEvents`, `MLEvaluation` and `CreateHistograms`.

### Current behavior

So far, the dynamic behavior results from the merging the outputs of `ReduceEvents` into files of a configurable size. The merging factor is computed by `MergeReductionStats` and therefore depends on various runtime decisions, such as initial file size, selection efficiency, amount of columns to be stored, etc. For this reason, the number of files that dictates the shape (i.e., the number of branches) of subsequent workflows cannot be known a priori and the task dependency tree can change in the course of actual task processing.

The stylized dependencies looked like this for the example of `UniteColumns`.

```mermaid
flowchart TD
    ... --> ReduceEvents
    ReduceEvents ==> MergeReduceEvents
    ReduceEvents ==> MergeReductionStats
    MergeReductionStats --> MergeReduceEvents
    MergeReduceEvents --> UniteColumns
```

### Updates

This PR brings several improvements:

- It is possible now to skip the merging step altogether. This allows to run all steps from calibration to selection, reduction and column coalescing in the same granularity, and even in `pilot` mode.
- The dynamic behavior is now implemented without custom caching and branch map resetting, but with generic law (`dynamic_workflow_condition`) and luigi (dynamic task dependencies) features.
- There is a new task, `ProvideReducedEvents`, that handles the switch between either requiring `ReduceEvents` with same branching, or `MergeReducedEvents` with changed branching (due to the merging). It can be configured to skip and force the merging, regardless of the computed merge factor. In the default case, the switch happens if the merge factor is 1 (no merging needed) or larger (merging required). It does not have outputs or an actual run method, but just triggers / forwards to the upstream task.
- The previous `MergedReducedEventsUser` interface, that mixed the dynamic behavior into the above mentioned tasks is refactored to a minimum and is consistently renamed to `ReducedEventsUser`.

The dependencies thus look like the following.

```mermaid
flowchart TD
    ... --> ReduceEvents
    ReduceEvents ==> MergeReducedEvents
    ReduceEvents ==> MergeReductionStats
    MergeReductionStats --> MergeReducedEvents
    ReduceEvents -.->|either| ProvideReducedEvents
    MergeReducedEvents -.->|or| ProvideReducedEvents
    ProvideReducedEvents --> UniteColumns
```

@mafrahm

### CODE REVIEW

**Comment by pkausw** [comment]
Hi @riga , any news here? I think the cross check questions are very minor, so after clarifying them this should be ready to merge

File: columnflow/tasks/reduction.py [hunk 1]
```diff
@@ -484,3 +456,131 @@ def merge(self, inputs, output):
...
+    )
+
+    @law.workflow_property(setter=True, cache=True, empty_value=0)
+    def file_merging(self):
+        if self.skip_merging:
+            return 1
+
+        # check if the merging stats are present
+        stats = self.reqs.MergeReductionStats.req(self).output()["stats"]
+        return stats.load(formatter="json")["merge_factor"] if stats.exists() else 0
```
[the remark this answers is not in the document]
Comment (riga) [reply columnflow/tasks/reduction.py:497]: Yup, correct :+1:

File: columnflow/tasks/histograms.py [hunk 2]
```diff
@@ -25,21 +25,19 @@ class CreateHistograms(
...
 ):
     sandbox = dev_sandbox(law.config.get("analysis", "default_columnar_sandbox"))
 
     # upstream requirements
     reqs = Requirements(
-        MergeReducedEventsUser.reqs,
+        ReducedEventsUser.reqs,
         RemoteWorkflow.reqs,
-        MergeReducedEvents=MergeReducedEvents,
+        ProvideReducedEvents=ProvideReducedEvents,
```
[the remark this answers is not in the document]
Comment (riga) [reply columnflow/tasks/histograms.py:40]: Indeed. It's provided already by extending `ReducedEventsUser.reqs` which they all do. Removed :+1:

File: columnflow/tasks/reduction.py [hunk 3]
```diff
@@ -309,33 +332,26 @@ def get_avg_std(values):
...
-        avg_size, std_size = get_avg_std(sizes)
-        std_size = (sum((s - avg_size)**2 for s in sizes) / n)**0.5
-        max_size_merged = self.merged_size * 1024**2
-        merge_factor = int(round(max_size_merged / avg_size))
-        merge_factor = min(max(1, merge_factor), self.dataset_info_inst.n_files)
-        n_merged = int(math.ceil(self.dataset_info_inst.n_files / merge_factor))
+        stats["n_test_files"] = n
+        stats["tot_size"] = sum(sizes)
+        stats["avg_size"], stats["std_size"] = get_avg_std(sizes)
+        stats["std_size"] = (sum((s - stats["avg_size"])**2 for s in sizes) / n)**0.5
```
[the remark this answers is not in the document]
Comment (riga) [reply columnflow/tasks/reduction.py:341]: Yup, duplicated by accident. Removed :+1:

File: columnflow/tasks/framework/base.py [hunk 4]
```diff
@@ -932,8 +932,10 @@ def file_merging_factor(self):
 
         if isinstance(self.file_merging, int):
             # interpret the file_merging attribute as the merging factor itself
-            # non-positive numbers mean "merge all in one"
-            n_merge = self.file_merging if self.file_merging > 0 else n_files
+            # zero means "merge all in one"
```
[the remark this answers is not in the document]
Comment (riga) [reply columnflow/tasks/framework/base.py:940]: Added :+1:

File: columnflow/tasks/ml.py [hunk 5]
```diff
@@ -667,7 +664,7 @@ def workflow_requires(self):
             producers=(self.producers,),
         )
 
-        reqs["events"] = self.reqs.MergeReducedEvents.req_different_branching(self)
+        reqs["events"] = self.reqs.ProvideReducedEvents.req(self, _exclude=self.exclude_params_branch)
```
[the remark this answers is not in the document]
Comment (riga) [reply columnflow/tasks/ml.py:669]: This was a stray left-over from a previous test. Good catch ...
Removed :+1:

**Comment by riga** [comment]
@pkausw Ok, this PR is finally back on track and ready to move on.

**Review by pkausw — APPROVED** [review]
LGTM, thanks @riga !

**Pull request merged**